File: scripts/workflow_state.py
"""Canonical workflow-state and transition validation for evidence controllers."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


ROOT = Path(__file__).resolve().parent.parent
CONTRACT_PATH = ROOT / "assets" / "workflow-state-contract-v1.json"


@lru_cache(maxsize=1)
def workflow_contract() -> dict[str, Any]:
    value = json.loads(CONTRACT_PATH.read_text("utf-8"))
    if value.get("schema_version") != "workflow-state-contract/1.0":
        raise ValueError("Workflow-state contract has the wrong schema version")
    if not isinstance(value.get("layers"), dict):
        raise ValueError("Workflow-state contract has no layer registry")
    journey = value.get("visible_journey")
    if (
        not isinstance(journey, dict)
        or journey.get("schema_version") != "visible-journey/1.0"
        or not isinstance(journey.get("milestones"), list)
        or len(journey["milestones"]) != 6
    ):
        raise ValueError(
            "Workflow-state contract has no canonical six-milestone visible journey"
        )
    milestone_ids = [item.get("id") for item in journey["milestones"]]
    if any(not item for item in milestone_ids) or len(set(milestone_ids)) != 6:
        raise ValueError("Visible journey milestone ids must be unique")
    for stage_id in (
        "inputs", "evidence_review", "decisions", "build_checks", "delivery"
    ):
        if stage_id not in journey.get("checkpoints", {}):
            raise ValueError(
                f"Visible journey omits controller checkpoint {stage_id}"
            )
    constitution = value.get("delivery_blocker_constitution")
    if not isinstance(constitution, dict) or constitution.get("schema_version") != "delivery-blocker-constitution/1.0":
        raise ValueError("Workflow-state contract has no delivery-blocker constitution")
    return value


def assert_delivery_blocker(
    blocked: bool,
    fatal_reason: str | None = None,
    *,
    domain: str | None = None,
) -> None:
    constitution = workflow_contract()["delivery_blocker_constitution"]
    if not blocked:
        if fatal_reason is not None:
            raise ValueError("A non-blocked delivery result cannot carry a fatal reason")
        return
    if domain in set(constitution.get("non_delivery_blocking_domains", [])):
        raise ValueError(f"{domain} is a degradation domain, not a delivery blocker")
    if fatal_reason not in constitution.get("fatal_reasons", {}):
        raise ValueError(
            f"Delivery block must name one declared fatal reason; received {fatal_reason!r}"
        )


def assert_state(
    layer: str,
    status: str,
    blocker_class: str | None,
    user_blocking: bool,
    *,
    stage: str | None = None,
) -> None:
    contract = workflow_contract()
    declaration = contract["layers"].get(layer)
    if not declaration:
        raise ValueError(f"Unknown workflow-state layer: {layer}")
    state = declaration.get("states", {}).get(status)
    if not state:
        raise ValueError(f"Unknown {layer} workflow state: {status}")
    allowed_blockers = state.get("blocker_classes")
    if allowed_blockers is not None and blocker_class not in allowed_blockers:
        raise ValueError(
            f"{layer}.{status} cannot own blocker class {blocker_class!r}"
        )
    expected_user_blocking = state.get("user_blocking")
    if expected_user_blocking is None:
        expected_user_blocking = blocker_class in set(contract["user_blocking_classes"])
    if user_blocking is not expected_user_blocking:
        raise ValueError(
            f"{layer}.{status} user_blocking={user_blocking!r} disagrees with blocker ownership"
        )
    allowed_stages = state.get("stages")
    if allowed_stages is not None and stage not in allowed_stages:
        raise ValueError(f"{layer}.{status} cannot occur at stage {stage!r}")


def assert_transition(
    layer: str,
    previous_status: str | None,
    next_status: str,
    *,
    reset: bool = False,
) -> None:
    declaration = workflow_contract()["layers"].get(layer)
    if not declaration:
        raise ValueError(f"Unknown workflow-state layer: {layer}")
    if next_status not in declaration.get("states", {}):
        raise ValueError(f"Unknown {layer} workflow state: {next_status}")
    if previous_status is None or reset:
        return
    if previous_status not in declaration.get("states", {}):
        raise ValueError(f"Unknown prior {layer} workflow state: {previous_status}")
    allowed = declaration.get("transitions", {}).get(previous_status, [])
    if next_status not in allowed:
        raise ValueError(
            f"Illegal {layer} workflow transition: {previous_status} -> {next_status}"
        )


# End of the shared workflow-state transition contract.

File: scripts/lib/workflow_state.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const CONTRACT_PATH = path.resolve(HERE, "../../assets/workflow-state-contract-v1.json");
export const WORKFLOW_STATE_CONTRACT = Object.freeze(
  JSON.parse(fs.readFileSync(CONTRACT_PATH, "utf8")),
);

if (WORKFLOW_STATE_CONTRACT.schema_version !== "workflow-state-contract/1.0") {
  throw new Error("Workflow-state contract has the wrong schema version");
}

export const VISIBLE_JOURNEY_CONTRACT = Object.freeze(
  structuredClone(WORKFLOW_STATE_CONTRACT.visible_journey),
);

if (
  VISIBLE_JOURNEY_CONTRACT?.schema_version !== "visible-journey/1.0" ||
  !Array.isArray(VISIBLE_JOURNEY_CONTRACT.milestones) ||
  VISIBLE_JOURNEY_CONTRACT.milestones.length !== 6
) {
  throw new Error("Workflow-state contract has no canonical six-milestone visible journey");
}
if (
  new Set(VISIBLE_JOURNEY_CONTRACT.milestones.map((item) => item.id)).size !==
  VISIBLE_JOURNEY_CONTRACT.milestones.length
) {
  throw new Error("Visible journey milestone ids must be unique");
}
for (const stageId of [
  "inputs",
  "evidence_review",
  "decisions",
  "build_checks",
  "review",
  "delivery",
]) {
  if (!VISIBLE_JOURNEY_CONTRACT.checkpoints?.[stageId]) {
    throw new Error(`Visible journey omits controller checkpoint ${stageId}`);
  }
}

const USER_BLOCKERS = new Set(WORKFLOW_STATE_CONTRACT.user_blocking_classes);
const DELIVERY_CONSTITUTION =
  WORKFLOW_STATE_CONTRACT.delivery_blocker_constitution;
const BLOCKED_OUTCOMES =
  DELIVERY_CONSTITUTION?.blocked_outcomes ?? {};
export const DELIVERY_BLOCKED_OUTCOMES = Object.freeze(
  structuredClone(BLOCKED_OUTCOMES),
);

if (
  DELIVERY_CONSTITUTION?.schema_version !==
  "delivery-blocker-constitution/1.0"
) {
  throw new Error("Workflow-state contract has no delivery-blocker constitution");
}

export const DELIVERY_FATAL_REASONS = Object.freeze(
  Object.keys(DELIVERY_CONSTITUTION.fatal_reasons ?? {}),
);

/**
 * Assert the narrow production delivery boundary.
 *
 * Internal controller work remains resumable and broker evidence degrades to
 * the forecast waterfall. A terminal delivery block is legal only when it is
 * assigned to one of the four declared, model-fatal reasons.
 */
export function assertDeliveryBlocker({
  blocked,
  fatalReason = null,
  domain = null,
} = {}) {
  const nonBlocking = new Set(
    DELIVERY_CONSTITUTION.non_delivery_blocking_domains ?? [],
  );
  if (blocked !== true) {
    if (fatalReason !== null) {
      throw new Error("A non-blocked delivery result cannot carry a fatal reason");
    }
    return true;
  }
  if (nonBlocking.has(domain)) {
    throw new Error(`${domain} is a degradation domain, not a delivery blocker`);
  }
  if (!DELIVERY_FATAL_REASONS.includes(fatalReason)) {
    throw new Error(
      `Delivery block must name one declared fatal reason; received ${JSON.stringify(fatalReason)}`,
    );
  }
  return true;
}

export function assertWorkflowState(
  layer,
  { status, blockerClass = null, userBlocking = false, stage = null },
) {
  const declaration = WORKFLOW_STATE_CONTRACT.layers?.[layer];
  if (!declaration) throw new Error(`Unknown workflow-state layer: ${layer}`);
  const state = declaration.states?.[status];
  if (!state) throw new Error(`Unknown ${layer} workflow state: ${status}`);
  if (
    Array.isArray(state.blocker_classes) &&
    !state.blocker_classes.includes(blockerClass)
  ) {
    throw new Error(
      `${layer}.${status} cannot own blocker class ${JSON.stringify(blockerClass)}`,
    );
  }
  if (Object.hasOwn(state, "user_blocking")) {
    const expectedUserBlocking =
      state.user_blocking === null
        ? USER_BLOCKERS.has(blockerClass)
        : state.user_blocking;
    if (userBlocking !== expectedUserBlocking) {
      throw new Error(
        `${layer}.${status} user_blocking=${JSON.stringify(userBlocking)} disagrees with blocker ownership`,
      );
    }
  }
  if (Array.isArray(state.stages) && !state.stages.includes(stage)) {
    throw new Error(`${layer}.${status} cannot occur at stage ${JSON.stringify(stage)}`);
  }
  return true;
}

export function assertWorkflowTransition(
  layer,
  previousStatus,
  nextStatus,
  { reset = false } = {},
) {
  const declaration = WORKFLOW_STATE_CONTRACT.layers?.[layer];
  if (!declaration) throw new Error(`Unknown workflow-state layer: ${layer}`);
  if (!declaration.states?.[nextStatus]) {
    throw new Error(`Unknown ${layer} workflow state: ${nextStatus}`);
  }
  if (previousStatus === null || previousStatus === undefined || reset) return true;
  if (!declaration.states?.[previousStatus]) {
    throw new Error(`Unknown prior ${layer} workflow state: ${previousStatus}`);
  }
  if (!(declaration.transitions?.[previousStatus] ?? []).includes(nextStatus)) {
    throw new Error(
      `Illegal ${layer} workflow transition: ${previousStatus} -> ${nextStatus}`,
    );
  }
  return true;
}

export function normaliseUserFlowResult(result) {
  if (!result || typeof result !== "object" || Array.isArray(result)) {
    throw new Error("User-flow result must be an object");
  }
  const value = { ...result };
  if (value.status === "ACTION_REQUIRED" && value.blocker_class === undefined) {
    value.blocker_class = "USER_DECISION";
  }
  if (value.status === "NEEDS_INTERNAL_WORK" && value.blocker_class === undefined) {
    value.blocker_class = "INTERNAL_WORK";
  }
  if (value.status === "BLOCKED" && value.blocker_class === undefined) {
    throw new Error(
      `BLOCKED user-flow result at ${value.stage ?? "unknown"} lacks typed blocker ownership`,
    );
  }
  if (value.status === "BLOCKED") {
    const binding = BLOCKED_OUTCOMES[value.outcome];
    if (!binding) {
      throw new Error(
        `BLOCKED user-flow outcome ${JSON.stringify(value.outcome)} is not declared by the delivery-blocker constitution`,
      );
    }
    if (
      value.fatal_reason !== undefined &&
      value.fatal_reason !== binding.fatal_reason
    ) {
      throw new Error(
        `BLOCKED user-flow outcome ${value.outcome} changed its declared fatal reason`,
      );
    }
    if (
      value.blocker_domain !== undefined &&
      value.blocker_domain !== binding.domain
    ) {
      throw new Error(
        `BLOCKED user-flow outcome ${value.outcome} changed its declared blocker domain`,
      );
    }
    value.fatal_reason = binding.fatal_reason;
    value.blocker_domain = binding.domain;
    assertDeliveryBlocker({
      blocked: true,
      fatalReason: value.fatal_reason,
      domain: value.blocker_domain,
    });
  }
  if (value.status === "NEEDS_INTERNAL_WORK") {
    if (value.blocker_class !== "INTERNAL_WORK") {
      throw new Error("NEEDS_INTERNAL_WORK must be owned by INTERNAL_WORK");
    }
    if (value.fatal_reason !== undefined || value.blocker_domain !== undefined) {
      throw new Error("NEEDS_INTERNAL_WORK is resumable internal work, not a fatal delivery blocker");
    }
  }
  if (["SCREEN", "PAUSED", "PASS_PENDING_MANUAL"].includes(value.status)) {
    if (value.blocker_class !== undefined && value.blocker_class !== null) {
      throw new Error(`${value.status} user-flow result cannot carry a blocker class`);
    }
    value.blocker_class = null;
    if (value.fatal_reason !== undefined || value.blocker_domain !== undefined) {
      throw new Error(`${value.status} user-flow result cannot carry a delivery blocker`);
    }
  }
  value.user_blocking = USER_BLOCKERS.has(value.blocker_class ?? null);
  assertWorkflowState("user_flow", {
    status: value.status,
    blockerClass: value.blocker_class ?? null,
    userBlocking: value.user_blocking,
    stage: value.stage,
  });
  return value;
}

export function assertPublicStateOwnership(value) {
  const status = value?.status ?? value?.response_status ?? null;
  const blockerClass = value?.blocker_class ?? value?.blockerClass ?? null;
  const userBlocking = value?.user_blocking ?? value?.userBlocking ?? null;
  if (status === "ACTION_REQUIRED" && !["USER_DECISION", "USER_EVIDENCE", "FATAL_SOURCE"].includes(blockerClass)) {
    throw new Error(`ACTION_REQUIRED cannot be owned by ${blockerClass ?? "no blocker class"}.`);
  }
  if (blockerClass === "INTERNAL_WORK" && userBlocking === true) {
    throw new Error("INTERNAL_WORK cannot be user-blocking.");
  }
  return value;
}
